libraries/AP_Math/control.cpp

```diff
@@ -324,6 +324,38 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
     shape_vel_accel_xy(vel_target, accel_input, vel, accel, accel_max, jerk_max, dt, limit_total_accel);
 }
 
+// limit_accel_xy limits the acceleration to prioritise acceleration perpendicular to the provided velocity vector.
+bool limit_accel_xy(const Vector2f& vel, Vector2f& accel, float accel_max)
+{
+    // check accel_max is defined
+    if (!is_positive(accel_max)) {
+        return false;
+    }
+    // limit acceleration to accel_max while prioritizing cross track acceleration
+    if (accel.length_squared() > sq(accel_max)) {
+        if (vel.is_zero()) {
```

Comment:
This is likely just being pedantic and I'm not sure on this one but `is.zero()` only compares each element. But we really  divide by the `length()`.

Should this be the following due to the `vec.normalize()`?

```suggestion
        if (is_zero(vel.length())) {
```
Or even more stringent ... 
```suggestion
        if (is_zero(vel.length_squared())) {
```

One other thought about `void Vector3F.normalize()`  
I'd be tempted to make a new `bool Vector3F.nomalized()` that does this division by zero check for us...

Looking through the code in some places we condition it upon `is_zero()`, `is_zero(vec.length())`, or `is_zero(vec.length())`

Basically some numerical consistency might be nice is our code. Thoughts?

Reply:
I would say this should always be:
XXX.is_zero()
And that function should be the written to ensure that we can safely normalise the vector without getting a divide by zero.

So if the best option is to use the sum of the squares of each element to ensure that then we should change .iszero() to do that.

That is what I would think is the best way to do this.

Reply:
I seem to recall a funny divide by zero that happened because while the vector was not zero (but very small), after converting it to a length it became zero. Maybe I am recalling incorrectly.

Reply:
Clearly, none of this holds back this PR. Just my numeric side likes consistency. 

@peterbarker thoughts on the this?

Changing `vec.is_zero()` is a far wider change and not something I would touch.  far more use cases and they are not consistent.

I'd be proposing every time we call `vec.normalize()` it would look like this because we do this anyways just with inconsistent methods using various metrics.
```cpp
if (!vec.normalize()) {
    /// vec length was zero
   // do stuff
}
```

instead of 
```cpp
if (!vec.is_zero() || ) {
    // vec length was zero
    // do stuff

} else {
    vec.normalize()
}
```

Reply:
I'm not sure we could change .is_zero() because we don't really know how the vector is going to be used later.  even just multiplying the vector by 0.5 could be enough to change it from !is_zero() to is_zero().

Reply:
> I seem to recall a funny divide by zero that happened because while the vector was not zero (but very small), after converting it to a length it became zero. Maybe I am recalling incorrectly.

There was a case of length_squared check in one place but not in another for scurves. Which is why i'm a little keen on being consistent.

Reply:
Using the equivalent of is_zero(lenght_squared)  instead of just checking each element is zero would just be a slightly more conservative test so I can't see that being a problem.

I would be happy to use .is_zero() or have .normalize() return a bool. Fixing .is_zero() to be a little more conservative using is_zero(lenght_squared) is probably less likely to trip up others in the future. I can see the attractiveness of if (vec.normalize()).

These are decisions I wait to get guidance from Tridge, Randy and Peter.

Reply:
This is probably something worth discussing at the next dev meeting.